src/core/ensemble_scheduler.cc

```diff
@@ -0,0 +1,531 @@
+
+    StepList res;
+    if (ensemble_status_.IsOk()) {
+      std::vector<size_t> updated_tensors;
+      ensemble_status_ = UpdateEnsembleState(completed_step, updated_tensors);
+      if (ensemble_status_.IsOk()) {
+        ensemble_status_ = GetNextSteps(updated_tensors, res);
+      }
+    }
+    // Error or no more progress (completed or deadlock)
+    if ((!ensemble_status_.IsOk()) || (inflight_request_counter_ == 0)) {
+      ensemble_status_ = FinishEnsemble();
+    } else {
+      ready_steps.swap(res);
+    }
+  }
+  return ensemble_status_;
+}
+
+Status
+EnsembleContext::UpdateEnsembleState(
+    const std::shared_ptr<Step>& completed_step,
+    std::vector<size_t>& updated_tensors)
+{
+  updated_tensors.clear();
+  if (completed_step == nullptr) {
+    for (size_t i = 0; i < tensor_data_.size(); i++) {
+      if (tensor_data_[i].second != nullptr) {
+        updated_tensors.push_back(i);
+      }
+    }
+  } else {
+    inflight_request_counter_--;
+    if (completed_step->request_status_.code() != RequestStatusCode::SUCCESS) {
+      return Status(
+          completed_step->request_status_.code(),
+          completed_step->request_status_.msg());
+    } else {
+      auto step_idx = completed_step->step_idx_;
+      RETURN_IF_ERROR(completed_step->response_provider_->FinalizeResponse(
+          *((*completed_step->backend_)())));
+      const auto& response_header =
+          completed_step->response_provider_->ResponseHeader();
+      for (const auto& output : response_header.output()) {
+        if (output.has_raw()) {
+          auto it =
+              info_->steps_[step_idx].output_to_tensor_.find(output.name());
+          if (it != info_->steps_[step_idx].output_to_tensor_.end()) {
+            auto& tensor_data = tensor_data_[it->second];
+            *(tensor_data.first.mutable_dims()) = output.raw().dims();
+            tensor_data.first.set_batch_byte_size(
+                output.raw().batch_byte_size());
+
+            RETURN_IF_ERROR(completed_step->response_provider_->GetSystemMemory(
+                it->first, &(tensor_data.second)));
+            updated_tensors.push_back(it->second);
+          } else {
+            return Status(
+                RequestStatusCode::INTERNAL,
+                "internal response header specified output '" + output.name() +
+                    "' that does not map to any ensemble tensors");
+          }
+        } else {
+          return Status(
+              RequestStatusCode::INTERNAL,
+              "internal response header should return output '" +
+                  output.name() +
+                  "' as raw data instead of classification result");
+        }
+      }
+    }
+  }
+  return Status::Success;
+}
+
+Status
+EnsembleContext::GetNextSteps(
+    const std::vector<size_t>& updated_tensors, StepList& steps)
+{
+  steps.clear();
+
+  std::set<size_t> next_step_idx;
+  // Get steps whose tensors used for input are set
```

Comment:
This is not a very efficient way to execute the graph. We could have a more direct represntation of the graph so we don't have to iterate over all the tensors. It is fine for a first implementation but at some point we may need to refine it.